tools/lua_sstable_consumer.cc

```diff
@@ -0,0 +1,1003 @@
+}
+
+template <typename T>
+int cxx_gc_l(lua_State* l) {
+    auto& v = pop_userdata_as_ref<T>(l, 1);
+    v.~T();
+    lua_pop(l, 2);
+    return 0;
+}
+
+const schema& get_schema_l(lua_State* l) {
+    lua_getglobal(l, "schema");
+    auto& s = pop_userdata_ref<const schema>(l, -1);
+    lua_pop(l, 1);
+    return s;
+}
+
+int sstable_index_l(lua_State* l) {
+    auto field = lua_tostring(l, 2);
+    auto& sst = pop_userdata_ref<const sstables::sstable>(l, 1);
+    lua_pop(l, 2);
+    if (!strcmp(field, "filename")) {
+        const auto f = sst.get_filename();
+        lua_pushstring(l, f.data());
+        return 1;
+    }
+    return 0;
+}
+
+int partition_start_index_l(lua_State* l) {
+    auto field = lua_tostring(l, 2);
+    auto& ps = pop_userdata_as_ref<partition_start>(l, 1);
+    lua_pop(l, 2);
+    auto& schema = get_schema_l(l);
+    if (!strcmp(field, "key")) {
+        const auto f = format("{}", ps.key().key().with_schema(schema));
+        lua_pushstring(l, f.data());
+        return 1;
+    } else if (!strcmp(field, "key_raw")) {
+        auto mb = managed_bytes(ps.key().key().representation());
+        push_userdata<managed_bytes>(l, std::move(mb), managed_bytes_metatable);
+        return 1;
+    } else if (!strcmp(field, "token")) {
+        lua_pushinteger(l, ps.key().token()._data);
+        return 1;
+    } else if (!strcmp(field, "tombstone")) {
+        if (!ps.partition_tombstone()) {
+            return 0;
+        }
+        push_userdata<tombstone>(l, ps.partition_tombstone(), tombstone_metatable);
+        return 1;
+    }
+    return 0;
+}
+
+int managed_bytes_new_l(lua_State* l) {
+    auto content = lua_tostring(l, 1);
+    bytes b;
+    try {
+        b = from_hex(content);
+    } catch (const std::exception& e) {
+        return luaL_error(l, "invalid_argument: %s", e.what());
+    }
+    lua_pop(l, 1);
+    push_userdata<managed_bytes>(l, managed_bytes(b), managed_bytes_metatable);
+    return 1;
+}
+
+int managed_bytes_to_hex_l(lua_State* l) {
+    auto& mb = pop_userdata_as_ref<managed_bytes>(l, 1);
+    auto hex = to_hex(to_bytes(mb));
+    lua_pop(l, 1);
+    lua_pushstring(l, hex.data());
+    return 1;
+}
+
+int managed_bytes_len_l(lua_State* l) {
+    auto& mb = pop_userdata_as_ref<managed_bytes>(l, 1);
+    lua_pop(l, 1);
+    lua_pushinteger(l, mb.size());
+    return 1;
+}
+
+int managed_bytes_eq_l(lua_State* l) {
+    auto& mb_a = pop_userdata_as_ref<managed_bytes>(l, 1);
+    auto& mb_b = pop_userdata_as_ref<managed_bytes>(l, 2);
+    const auto eq = mb_a == mb_b;
+    lua_pop(l, 2);
+    lua_pushboolean(l, eq);
+    return 1;
+}
+
+void push_gc_clock_time_point(lua_State* l, gc_clock::time_point tp) {
+    auto t = gc_clock::to_time_t(tp);
+    std::tm tm{};
+    // This time is actually UTC, but C API assumes local-times and there is no
+    // way to specify the TZ used. So we lie about it being local-time to avoid
+    // libC doing TZ adjustments.
+    localtime_r(&t, &tm);
```

Comment:
Maybe the stuff here is helpful? https://en.cppreference.com/w/cpp/chrono

Reply:
This was the first place I looked for help. Sadly `std::chrono` is still incomplete, there is no proper way to parse timestamps, see 5ea6700e23c9b57f770677d272feb592e77d4c39.
In any case, when looking for how to represent time points in Lua I found this: http://www.lua.org/manual/5.4/manual.html#pdf-os.time
This describes a table with fields matching that of C's https://en.cppreference.com/w/c/chrono/tm, so I decided to use the same format, which kind-of brought with itself the usage of the C counterpart as well.